include/circt/Dialect/FIRRTL/FIRRTLModuleContext.h

```diff
@@ -0,0 +1,173 @@
+//===- FIRRTLModuleContext.h - Module context base --------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Provides context information that is global to the module.
+//
+// The pure virtual class `FIRRTLModuleContext` is inherited and implemented in
+// `FIRParser.cpp` and `FIRRTLFFIContext.cpp`.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_DIALECT_FIRRTL_FIRRTLMODULECONTEXT_H
+#define CIRCT_DIALECT_FIRRTL_FIRRTLMODULECONTEXT_H
+
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/IR/Value.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/PointerEmbeddedInt.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/Twine.h"
+#include "llvm/Support/Allocator.h"
+#include <string>
+
+namespace circt {
+namespace firrtl {
+
+//===----------------------------------------------------------------------===//
+// FIRRTLModuleContext
+//===----------------------------------------------------------------------===//
+
+class FIRRTLModuleContext {
+public:
+  // Entries in a symbol table are either an mlir::Value for the operation that
+  // defines the value or an unbundled ID tracking the index in the
+  // UnbundledValues list.
+  using UnbundledID = llvm::PointerEmbeddedInt<unsigned, 31>;
+  using SymbolValueEntry = llvm::PointerUnion<Value, UnbundledID>;
+
+  using ModuleSymbolTable =
+      llvm::StringMap<std::pair<llvm::SMLoc, SymbolValueEntry>,
+                      llvm::BumpPtrAllocator>;
+  using ModuleSymbolTableEntry =
+      llvm::StringMapEntry<std::pair<llvm::SMLoc, SymbolValueEntry>>;
+
+  using UnbundledValueEntry = SmallVector<std::pair<Attribute, Value>>;
+  using UnbundledValuesList = std::vector<UnbundledValueEntry>;
```

Comment:
SmallVector generally preferred for uniformity and compatibility with ranges and such.